synapse/lib/auth.py

```diff
@@ -1332,7 +1339,25 @@ async def tryPasswd(self, passwd, nexs=True):
             return False
 
         if isinstance(shadow, dict):
-            return await s_passwd.checkShadowV2(passwd=passwd, shadow=shadow)
+            result = await s_passwd.checkShadowV2(passwd=passwd, shadow=shadow)
+            if self.auth.policy and (attempts := self.auth.policy.get('attempts')) is not None:
+                if result:
+                    await self.auth.setUserInfo(self.iden, 'policy:attempts', 0)
+                    return True
+
+                valu = self.info.get('policy:attempts', 0) + 1
+                await self.auth.setUserInfo(self.iden, 'policy:attempts', valu)
+
+                if valu >= attempts:
+                    await self.auth.nexsroot.cell.setUserLocked(self.iden, True)
+
+                    mesg = f'User {self.name} has exceeded the number of allowed password attempts ({valu +1}), locking their account.'
+                    extra = {'synapse': {'target_user': self.iden, 'target_username': self.name, 'status': 'MODIFY'}}
+                    logger.info(mesg, extra=extra)
+
+                    raise s_exc.AuthDeny(mesg=mesg, attempts=attempts, user=self.iden, username=self.name)
```

Comment:
We should discuss this returning `False` here. There is a odd pattern here where we **only** raise when we do the automatic lock, and do not raise on a subsequent failure that was caused by a lockout that happened as a result of the automatic lock.